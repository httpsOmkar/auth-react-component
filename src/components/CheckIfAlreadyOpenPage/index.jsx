import React from 'react';
import { element } from 'prop-types';

/**
 * This component for reopening the same page in two or more tabs.
 * To avoid the token errors
 */
export default class CheckIfAlreadyOpenPage extends React.Component {
  static propTypes = {
    /**
     * Render children when the page is already open
     */
    errorChildren: element.isRequired,

    /**
     * Children when there is no error
     */
    children: element.isRequired,
  };

  static defaultProps = {
    errorChildren: null,
    children: null,
  };

  state = {
    ready: false,
    alreadyOpen: false,
  };

  componentDidMount() {
    if (localStorage && window) {
      localStorage.setItem('page_opened', Date.now().toString());
      const onLocalStorageEvent = event => {
        if (event.key === 'page_opened') {
          // Emit that you're already available.
          return localStorage.setItem('already_open', Date.now().toString());
        }

        if (event.key === 'already_open') {
          return this.setState({ alreadyOpen: true });
        }

        this.setState({ ready: true });
      };

      window.addEventListener('storage', onLocalStorageEvent, false);
    }
  }

  render() {
    const { alreadyOpen, ready } = this.state;
    const { errorChildren, children } = this.props;

    if (ready) {
      return alreadyOpen ? errorChildren : children;
    }

    return <></>;
  }
}
