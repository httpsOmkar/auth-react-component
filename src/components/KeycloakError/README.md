```js
const { KeycloakContext } = require('../KeycloakWrapper');

<div>
    <KeycloakContext.Provider value={{ error: 'Something went wrong' }}>
      <KeycloakError/>
    </KeycloakContext.Provider>
    <KeycloakContext.Provider value={{ error: new Error('This is error occurred in keycloak') }}>
      <KeycloakError/>
    </KeycloakContext.Provider>
</div>

```
