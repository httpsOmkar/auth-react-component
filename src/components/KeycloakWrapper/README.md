```js
const KeycloakInstance = require('../../utils/keycloak-mock-instance').default;

const kc = new KeycloakInstance();

<KeycloakWrapper keycloak={kc} keycloakOptions={ {  } } tokenUpdateInterval={100}>
  <p>I am protected</p>
</KeycloakWrapper>

```

### Failing error

```js
const KeycloakInstance = require('../../utils/keycloak-mock-instance').default;

const kc = new KeycloakInstance();
kc.isError = true;
<KeycloakWrapper keycloak={kc} keycloakOptions={ {  } } tokenUpdateInterval={100}>
  <p>I am protected</p>
</KeycloakWrapper>

```
