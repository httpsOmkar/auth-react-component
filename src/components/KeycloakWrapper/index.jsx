import React, { Component } from 'react';
import { string, element, any, object, number } from 'prop-types';
import KeycloakError from '../KeycloakError';

export const KeycloakContext = React.createContext();

/**
 * Render children when authenticated with the keycloak.
 */
export default class KeycloakWrapper extends Component {
  static propTypes = {
    /** The CSS class name of the wrapper element */
    className: string,

    /**
     * Element which will render when the keycloak is initializing
     */
    loadingElement: element.isRequired,

    /**
     * Child element to render
     */
    children: element.isRequired,

    /**
     * Keycloak instance
     */
    keycloak: any.isRequired,

    /**
     * Options for keycloak
     */
    keycloakOptions: object.isRequired,

    /**
     * Render when any error is there
     */
    errorChildren: element,

    /**
     * Update token in this interval
     */
    tokenUpdateInterval: number.isRequired,
  };

  static defaultProps = {
    className: null,
    loadingElement: <p>Loading...</p>,
    children: null,
    keycloak: null,
    tokenUpdateInterval: null,
    errorChildren: null,
  };

  state = {
    keycloakPromise: null,
    error: null,
  };

  /**
   * @private
   * Interval id
   */
  refreshIntervalId;

  /**
   * @private
   * Keycloak promise
   * @returns {Promise<any>}
   */
  keycloakPromise() {
    const { keycloak } = this.props;

    return new Promise((resolve, reject) => {
      // Call the init keycloak
      keycloak
        .init(this.props.keycloakOptions)
        .success(authenticated => {
          if (authenticated) {
            // Update the token in specific time
            this.refreshIntervalId = setInterval(() => {
              keycloak.updateToken(0).error(err => {
                if (err) {
                  // eslint-disable-next-line
                  this.setState({ error: err });
                }
              });
            }, this.props.tokenUpdateInterval);
            resolve(authenticated);
          } else {
            keycloak.login();
          }
        })
        .error(reject);
    });
  }

  /**
   * @private
   * Load the keycloak profile
   * @returns {Promise<any>}
   */
  keycloakProfilePromise() {
    const { keycloak } = this.props;

    return new Promise((resolve, reject) => {
      keycloak
        .loadUserProfile()
        .success(resolve)
        .error(reject);
    });
  }

  componentDidMount() {
    this.keycloakPromise()
      .then(docs => {
        this.setState({ keycloakPromise: docs });
      })
      .catch(error => {
        this.setState({ error });
      });
  }

  componentWillUnmount() {
    // Clear the interval if exits
    if (this.refreshIntervalId) {
      clearInterval(this.refreshIntervalId);
    }
  }

  /**
   * Render the element
   * @returns {*}
   */
  render() {
    const { keycloakPromise, error } = this.state;
    const { children, errorChildren } = this.props;
    const ErrorChild = errorChildren || KeycloakError;

    return (
      <KeycloakContext.Provider value={this.state}>
        <>
          {!error && <>{keycloakPromise && children}</>}
          {error && <ErrorChild />}
        </>
      </KeycloakContext.Provider>
    );
  }
}
