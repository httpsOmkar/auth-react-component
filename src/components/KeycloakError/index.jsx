import React, { PureComponent } from 'react';
import { string } from 'prop-types';
import { KeycloakContext } from '../KeycloakWrapper';

/**
 * Render the Keycloak ERROR if there any
 */
export default class KeycloakError extends PureComponent {
  static propTypes = {
    /** The CSS class name of the wrapper element */
    className: string,
  };

  /**
   * Normalize the error into human readable string
   * @param error
   * @returns {String|null}
   */
  static normalizeError(error) {
    if (!error) {
      return null;
    }

    return error.message || error;
  }

  /**
   * Render the elements
   * @returns {*}
   */
  render() {
    return (
      <KeycloakContext.Consumer>
        {({ error }) => (
          <div className={this.props.className || ''}>
            {KeycloakError.normalizeError(error)}
          </div>
        )}
      </KeycloakContext.Consumer>
    );
  }
}
