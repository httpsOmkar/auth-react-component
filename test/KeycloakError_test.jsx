import React from 'react';
import { mount } from 'enzyme';
import KeycloakError from '../src/components/KeycloakError';
import { KeycloakContext } from '../src/components/KeycloakWrapper';

describe('<KeycloakWrapper />', () => {
  it('should render correctly', () => {
    const error = mount(
      <KeycloakContext.Provider value={{ error: 'Something went wrong' }}>
        <KeycloakError />
      </KeycloakContext.Provider>
    );

    expect(error).toMatchSnapshot();
  });

  it('should render not render when no error', () => {
    const error = mount(
      <KeycloakContext.Provider value={{ error: null }}>
        <KeycloakError />
      </KeycloakContext.Provider>
    );

    expect(error).toMatchSnapshot();
  });
});
