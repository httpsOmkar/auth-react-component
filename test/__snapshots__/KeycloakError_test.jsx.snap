// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeycloakWrapper /> should render correctly 1`] = `
<KeycloakError>
  <div
    className=""
  >
    Something went wrong
  </div>
</KeycloakError>
`;

exports[`<KeycloakWrapper /> should render not render when no error 1`] = `
<KeycloakError>
  <div
    className=""
  />
</KeycloakError>
`;
